cmd/neofs-node/policy.go

```diff
@@ -2,26 +2,113 @@ package main
 
 import (
 	"fmt"
+	"sync"
 
+	"github.com/hashicorp/golang-lru/v2/simplelru"
 	"github.com/nspcc-dev/neofs-node/pkg/core/container"
 	"github.com/nspcc-dev/neofs-node/pkg/core/netmap"
 	cid "github.com/nspcc-dev/neofs-sdk-go/container/id"
 	netmapsdk "github.com/nspcc-dev/neofs-sdk-go/netmap"
 )
 
+// storagePolicyRes structures persistent storage policy application result for
+// particular container and network map incl. error.
+type storagePolicyRes struct {
+	nodeSets [][]netmapsdk.NodeInfo
+	err      error
+}
+
+// containerNodesAtEpoch is a thread-safe LRU cache mapping containers into
+// storage policy application results for particular network map.
+type containerNodesAtEpoch struct {
+	mtx sync.RWMutex
+	lru simplelru.LRUCache[cid.ID, storagePolicyRes]
+}
+
+// containerNodesCaches groups caches for all tracked epochs.
+type containerNodesCaches struct{ cur, prev *containerNodesAtEpoch }
+
+// max number of container storage policy applications results cached by for
+// each epoch.
+const cachedContainerNodesPerEpochNum = 1000
+
 // containerNodes wraps NeoFS network state to apply container storage policies.
+//
+// Since policy application results are consistent for fixed container and
+// network map, they could be cached. The containerNodes caches up to
+// cachedContainerNodesPerEpochNum results for the latest and the previous
+// epochs. The previous one is required to support data migration after the
+// epoch tick. Older epochs are not cached.
 type containerNodes struct {
 	containers container.Source
 	network    netmap.Source
+
+	lastSeenCurrentEpochMtx sync.Mutex
+	lastSeenCurrentEpoch    uint64
+	epochCaches             containerNodesCaches
```

Comment:
And maybe even simplify these dances.

Reply:
i dont find them complex. U may leave an issue for the future

Reply:
redone